core/drivers/ns16550.c

```diff
@@ -42,16 +45,46 @@
 /* uart status register bits */
 #define UART_LSR_THRE	0x20 /* Transmit-hold-register empty */
 
-void ns16550_flush(vaddr_t base)
+static vaddr_t base_addr(struct ns16550_data *pd)
 {
+	return cpu_mmu_enabled() ? pd->vbase : pd->pbase;
+}
+
+static void ns16550_flush(struct serial_chip *chip)
+{
+	struct ns16550_data *pd = container_of(chip, struct ns16550_data, chip);
+	vaddr_t base = base_addr(pd);
+
 	while ((read8(base + UART_LSR) & UART_LSR_THRE) == 0)
 		;
 }
 
-void ns16550_putc(int ch, vaddr_t base)
+static void ns16550_putc(struct serial_chip *chip, int ch)
 {
-	ns16550_flush(base);
+	struct ns16550_data *pd = container_of(chip, struct ns16550_data, chip);
+	vaddr_t base = base_addr(pd);
+
+	ns16550_flush(chip);
 
 	/* write out charset to Transmit-hold-register */
 	write8(ch, base + UART_THR);
 }
+
+static const struct serial_ops ns16550_ops = {
+	.flush = ns16550_flush,
+	.putc = ns16550_putc,
+};
+
+void ns16550_init(struct ns16550_data *pd, vaddr_t base)
+{
+	if (cpu_mmu_enabled())
```

Comment:
`ns16550_init()` is always called with argument `base` set to `CONSOLE_UART_BASE`, not the vaddr.